modules/helper/EsxiServiceHelper.py

```diff
@@ -0,0 +1,106 @@
+from interfaces.host import Host
+from modules.api.Atlas import Atlas
+from modules.api.SshHelper import SshHelper
+
+from threading import Thread
+from queue import Queue
+from typing import List
+
+import logging
+
+logger = logging.getLogger('esxi')
+
+
+class EsxiServiceHelper:
+
+    def __init__(self, atlas: Atlas, esxi_username: str, esxi_password: str, monitored_services, max_threads: int):
+        """
+        Helper providing functionality to query critical service stats of esxi-host
+        """
+
+        self.atlas = atlas
+        self.esxi_username = esxi_username
+        self.esxi_password = esxi_password
+        self.max_threads = max_threads
+
+        self._services = monitored_services
+        command_list = ["/etc/init.d/%s status" %
+                        service for service in self._services]
+        self._command = ' & '.join(command_list)
+
+        # Strange bug. Display name is different from registered service name
+        self._services = [item.replace('nsx-opsagent', 'opsAgent')
+                          for item in self._services]
+
+    def _worker(host_queue: Queue, ssh_username: str, ssh_password: str,
+                command: str, services: list, output: list):
+        """
+        Multithreaded worker. Used for SSH-Connections to esxi-hosts
+        """
+
+        while not host_queue.empty():
+            host: Host = host_queue.get()
+
+            answer = SshHelper.execute_command(
+                host.address,
+                ssh_username,
+                ssh_password,
+                command
+            )
+
+            host.services = {}
+            if answer is None:
+                continue
+
+            # loop over all critical services and try to find them in the answer
+            answer = answer.splitlines()
+            for service in services:
+                for line in answer:
+                    if service in line and 'is running' in line:
+                        host.services[service] = True
+                        # check next service
+                        break
+                    else:
+                        host.services[service] = False
+            output.append(host)
+
+    def get_all_service_stats(self) -> List[Host]:
+        """
+        Get all critical service stats of all listed esxi-hosts.
+
+        :return: List of Host
+        """
+
+        # get hosts
+        hosts = self.atlas.get_esxi_hosts()
+
+        # get services
+        host_queue = Queue()
+        [host_queue.put(host) for host in hosts]
```

Comment:
Interesting approach of tackling this. However, you are limiting threads to open a defined number of connections. This could be done with a Semaphore. Create it with the count of 10 and have your threads acquiring the resources. The Semaphore object will then handle the execution of these threads. 

The fact that we need that big of a comment to understand it, indicates, there might be another solution :).